### PR TITLE
feature: refresh the page every 30 seconds

### PR DESCRIPTION
**Pull request by Bill-1991**
# Description
refresh the page and content every 30 secs

Fixes # 87

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Chore items (this includes basic clean up of files, or updates to packages)
- [ ] Updates to documentation

## Checklist

- [x] I have a descriptive title for my PR
- [x] I have linked the issue to this [PR](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] I have run the project locally and reviewed the code changes
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by dgupta0** [comment]
I did these changes inside` dom` event listener and put the fetch call inside a `getData` function. 

I am adding a reference image too. I am not sure if I have to create a new PR so I am leaving a comment for what I have come up with. 

Screenshot 2023-03-25 at 8 32 17 AM

```
getData()
  // Fetch FCC forum latest data
```
 ```
 setInterval(() => {
    categoryBtns.innerHTML = ""
    postsContainer.innerHTML = ""
    userListContainer.innerHTML = ""
    isLoading = true;
    getData()
  }, 30000)
});
```

**Comment by Dorfieeee** [comment]
Hey @dgupta0 !

This may work but we are trying to keep the old posts while adding new ones so by wiping out all the HTML we would loose them. The goal is to update the existing posts with new data(view count, title name, ...) and add new ones. Additionally, when the user has selected to display only certain category and has sorted the posts by some property, it should remain that way after the update. That means that the way user is seeing the forum should not be affected by the update.

**Comment by dgupta0** [comment]
hhhmm, not sure how much I understand what we are trying to do here then.  I think what I had in mind was, if we just make a fetch call after 30 seconds and clear the html, then we will automatically update the page with all the latest content. Though while I was writing the code for refresh every 30 seconds, I randomly did lots of changes and this was something I thought was working.  Thanks for getting back :)

**Comment by Dorfieeee** [comment]
We want to retain the existing posts(the ones already in the table) so the table will have more than 30 posts over time depending how long you have the forum app opened. We also want to update the posts with the most recent details. That's why we cannot just replace the table with the most recent posts.

**Comment by dgupta0** [comment]
Okay so we want to add more posts and update details if any after the time interval. And the solution I am suggesting will update the existing posts details but because I am clearing the inner html will not add more recent posts to the already displaying 30 posts but render 30 (new) recent posts. I hope I am explaining it correctly.

**Review by Dorfieeee — COMMENTED** [review]
1. That code repetition in `handleSortBtnClick` & `handleSortTopicsOnRefresh` is certainly not a good practice.

`handleSortBtnClick` could only change the `sortBy.state` & `sortBy.order` and then call the sorting function and display the posts.

> simplified example
```
function handleSortBtnClick() {
sortBy.state = "state"
sortBy.order = "order"
sortTopics()
displayTopics()
}
```

2. I still don't really understand, why do we need separate mechanisms for sorting, filtering, page refresh... Why cannot we just use `refreshPage()` for the initial load and all the others as well? The same applies for the other mechanisms.

Build functions, which ideally do only one thing and compose them together.

One way to have only one function for each mechanisms is to include them inside the render method.
image

File: index.js [hunk 1]
```diff
@@ -144,40 +158,33 @@ function displayCategories() {
   });
 }
 
+function filterTopics(topics, filterBy) {
```
[the remark this answers is not in the document]
Comment (Bill-1991) [reply index.js:161]: I was sure you'd point that out as i thought of it after i left it there :p...now comes the hard part for combined topics and users

[the remark this answers is not in the document]
Comment (Bill-1991) [reply index.js:161]: after the return statement in if i have to return topics cause it is gonna leave it as undefined

[the remark this answers is not in the document]
Comment (Bill-1991) [reply index.js:161]: no,that's not correct either...what do you mean by !filterBy here?...if it's not filterBy?

[the remark this answers is not in the document]
Comment (Bill-1991) [reply index.js:161]: if (!filterBy) return topics;
  let categoryId = parseInt(filterBy);
  let filteredTopics = topics.filter(topic => topic.category_id === categoryId);
  return filteredTopics;

That only works for me
I'm gonna leave it there and come back to it

[the remark this answers is not in the document]
Comment (Dorfieeee) [reply index.js:161]: !string is the same as string === "" or string.lenght === 0

[the remark this answers is not in the document]
Comment (Dorfieeee) [reply index.js:161]: https://developer.mozilla.org/en-US/docs/Glossary/Truthy

**Pull request merged**